### PR TITLE
Mix N' Match exposure times

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR enabled Mix N' Match to have different exposure times for different filters.

### CODE REVIEW

**Comment by scizen9** [comment]
Could someone post an example payload with the individual exptimes?  It's hard to parse the commit and anticipate what the sedm watcher script will be receiving.  Will this only work for Mix 'n Match?  It might be worth implementing a scheme that would work for all modes.  I can imagine it would be useful to specify separate exposure times for IFU and RC filter images.  The sedm watcher treats the IFU like another filter, so you could just either send filter/exptime pairs or an overarching exptime that would be applied to all filters, or the default -1, in which case the watcher will calculate exptime for each filter based on recent photometry.

**Comment by mcoughlin** [comment]
@scizen9 

{'Filters': 'u,g,r,i', 'Followup': '', 'email': 'test_suite@skyportal.com', 'enddate': '2023-06-24', 'startdate': '2023-06-17', 'prior_photometry': {}, 'priority': 1, 'programname': 'Program A', 'requestid': 11, 'sourceid': 'RedbackMillisecondPulsar', 'sourcename': 'RedbackMillisecondPulsar', 'status': 'new', 'username': 'provisioned-admin', 'ra': 33.043637, 'dec': 53.36078, 'exptime': [-1, 340, 30, 300], 'maxairmass': 2.8, 'max_fwhm': 10}

I added the same change for the other options.

File: skyportal/facility_apis/sedm.py [hunk 1]
```diff
@@ -392,10 +430,8 @@ def prepare_payload(payload, existing_payload=None):
                             "advanced": {"enum": [True]},
                             "exposure_time": {
                                 "title": "Exposure Time (Photometry) [s]",
-                                "type": "number",
-                                "default": -1,
-                                "minimum": -1,
-                                "maximum": 3600,
```
Comment (Theodlz) [inline skyportal/facility_apis/sedm.py:398]: I believe Don had suggested this maximum before. We can remove it here, but if @scizen9 wants to keep that limit (or in general if we want to avoid number lower than -1), we should add some verification somewhere right after the one you added line 184.

File: static/js/components/FollowupRequestForm.jsx [hunk 2]
```diff
@@ -169,12 +169,62 @@ const FollowupRequestForm = ({
...
+    if (formData?.advanced) {
+      if (formData?.exposure_time) {
+        let exposure_times = [];
+        if (typeof formData.exposure_time === "string") {
+          exposure_times = formData.exposure_time.split(",").map(Number);
+        } else {
+          exposure_times = formData.exposure_time;
+        }
+        for (let i = 0; i < exposure_times.length; i += 1) {
+          if (exposure_times[i] < -1 || exposure_times[i] > 3600) {
```
Comment (Theodlz) [inline static/js/components/FollowupRequestForm.jsx:195]: Nice. Let's have something similar is the backend for people that use this API

**Review by Theodlz — COMMENTED** [review]
I left a quick review @mcoughlin.

File: skyportal/facility_apis/sedm.py [hunk 3]
```diff
@@ -172,6 +172,44 @@ def prepare_payload_sedm(payload, existing_payload=None):
...
+    }
+
+    if type(payload["exposure_time"]) == str:
+        payload["exposure_time"] = [
+            int(exposure_time) for exposure_time in payload["exposure_time"].split(",")
+        ]
+
+    if payload['observation_type'] == "Mix 'n Match":
+        # All filters have same exposure time
+        if len(payload["exposure_time"]) == 1:
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/facility_apis/sedm.py:191]: good catch!

File: skyportal/facility_apis/sedm.py [hunk 4]
```diff
@@ -172,6 +172,44 @@ def prepare_payload_sedm(payload, existing_payload=None):
...
+            payload["exposure_time"] = payload["exposure_time"][0]
+        else:
+            # Confirm that number of exposure times matches number of
+            # observation choices
+            if not len(payload["exposure_time"]) == len(payload['observation_choices']):
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/facility_apis/sedm.py:196]: ok

File: skyportal/facility_apis/sedm.py [hunk 5]
```diff
@@ -172,6 +172,44 @@ def prepare_payload_sedm(payload, existing_payload=None):
...
+                raise ValueError(
+                    "Exposure times should either have one entry (same exposure time for all filters) or the same length as observation_choices"
+                )
+    else:
+        num_exposures = num_obs[payload['observation_type']]
+        # All filters have same exposure time
+        if len(payload["exposure_time"]) == 1:
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/facility_apis/sedm.py:203]: oops

File: skyportal/facility_apis/sedm.py [hunk 6]
```diff
@@ -172,6 +172,44 @@ def prepare_payload_sedm(payload, existing_payload=None):
...
+            payload["exposure_time"] = payload["exposure_time"][0]
+        else:
+            # Confirm that number of exposure times matches number of
+            # filters
+            if not len(payload["exposure_time"]) == num_exposures:
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/facility_apis/sedm.py:208]: oooppppssss

File: skyportal/facility_apis/sedm.py (hunk 1 above)
Comment (mcoughlin) [reply skyportal/facility_apis/sedm.py:398]: Added!

**Comment by scizen9** [comment]
Would it be possible/practical to always give the filters and exptimes as a list?   The exptimes list could be filled with -1s if no exptimes are specified.  Also, how does the front-end receive exposure times for each filter?  In the advanced tab, do you have an exptime entry box for each filter?

File: skyportal/facility_apis/sedm.py (hunk 1 above)
Comment (scizen9) [reply skyportal/facility_apis/sedm.py:398]: Currently, the sedm watcher code just checks if exptime < 0, so any negative number will work.

**Comment by Theodlz** [comment]
> Would it be possible/practical to always give the filters and exptimes as a list?   The exptimes list could be filled with -1s if no exptimes are specified.  Also, how does the front-end receive exposure times for each filter?  In the advanced tab, do you have an exptime entry box for each filter?

Sure! On the frontend they can just be specified as a list, just like for the filters (using commas). The frontend input is then converted to a list for the exptime in the backend. There's some error handling to show the user an error if he for example gave a list of 2 exptimes but 3 filters, so that should be pretty robust.

When you say that you'd like the filters as a list, do you mean not as a comma seperated list (which is a string) but as a properly list? If yes, then that shouldnt be a problem.

I guess my question is: if the user asks for just one filter, or many filters but just one exposure time, do you still always want to receive everything as proper lists and not single values or comma seperated string lists?

Let me know if that makes sense @scizen9

**Comment by mcoughlin** [comment]
@scizen9 @Theodlz I modified the code to always return a list.

**Comment by scizen9** [comment]
Looking at the code and PR commits, I am thinking that we should send the filter and exptime lists as comma separated strings (exptimes would be integer string values).  That is how the SEDM watcher scripts expects the filter list.

It seems that the only time you send a string in the 'Followup' field is when an IFU observation is requested, then it sends 'IFU' in the 'Followup' item.  Do you plan to continue this procedure with the mix 'n match if an IFU is requested?

The easiest thing for me to handle is to have a one-to-one correspondence between filters in the list and exptimes in the list and to also include 'ifu', if so requested.  That way, I can basically ignore the 'Followup' item and just parse the two lists to generate the exptimes for each observation (filters and ifu).

We probably always want the IFU observation to be first, so re-ordering the request to always put the ifu first in the list (if requested) would be very helpful.

**Comment by mcoughlin** [comment]
@scizen9 Please see the updated code, which should reflect your desired ordering and the comma delimited strings.

**Comment by Theodlz** [comment]
@mcoughlin I would maybe close this one just like we did with the other. If this becomes important again, we can just re-open
